javaslang/src/main/java/javaslang/collection/Vector.java

```diff
@@ -11,39 +11,41 @@
 
 import java.io.Serializable;
 import java.util.*;
-import java.util.Set;
 import java.util.function.*;
 import java.util.stream.Collector;
 
+import static javaslang.collection.BitMappedTrie.lastDigit;
 import static javaslang.collection.Collections.*;
 
 /**
- * Vector is the default Seq implementation. It provides the best performance in between Array (with constant time element access)
- * and List (with constant time element addition).
+ * Vector is the default Seq implementation that provides effectively constant time access to any element.
+ * Many other operations (e.g. `tail`, `drop`, `slice`) are also effectively constant.
  *
  * @param <T> Component type of the Vector.
- * @author Ruslan Sennov
+ * @author Ruslan Sennov, Pap Lőrinc
+ * @see BitMappedTrie
  * @since 2.0.0
  */
 public final class Vector<T> implements Kind1<Vector<?>, T>, IndexedSeq<T>, Serializable {
     private static final long serialVersionUID = 1L;
 
-    private static final Vector<?> EMPTY = new Vector<>(HashArrayMappedTrie.empty());
+    private static final Vector<?> EMPTY = new Vector<>(BitMappedTrie.empty());
 
-    private final HashArrayMappedTrie<Integer, T> trie;
-    private final int indexShift;
-
-    private static <T> Vector<T> wrap(HashArrayMappedTrie<Integer, T> trie) {
-        return trie.isEmpty() ? empty() : new Vector<>(trie);
+    final BitMappedTrie<T> trie;
+    private Vector(BitMappedTrie<T> trie) {
+        assert (EMPTY == null) || (trie.length() > 0);
```

Comment:
Wow, this is tricky :-) During initialization of EMPTY the constructor is called and EMPTY == null is true.

In all other calls of the constructor we assert that the empty trie is not passed to ensure the EMPTY Vector is a singleton.

I had to proof it to understand the assert statement:

```java
    (EMPTY == null) || (trie.length() > 0)   // applying De Morgan
<=> !((EMPTY != null) && (trie.length <= 0)) // my assumption: (EMPTY != null) is true
<=> !(true && (trie.length <= 0))            // true && A <=> A
<=> !(trie.length <= 0)                      // negation
<=> trie.length > 0                          // not correct for EMPTY Vector
```